pkg/operations/kubernetesupgrade/upgrader.go

```diff
@@ -216,51 +215,86 @@ func (ku *Upgrader) upgradeAgentPools() error {
 		upgradeAgentNode.Client = ku.Client
 		upgradeAgentNode.kubeConfig = ku.kubeConfig
 
-		upgradedAgentsIndex := make(map[int]bool)
+		upgradeVMs := make(map[int]*UpgradeVM)
 
 		for _, vm := range *agentPool.UpgradedAgentVMs {
 			ku.logger.Infof("Agent VM: %s, pool name: %s on expected orchestrator version\n", *vm.Name, *agentPool.Name)
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
-			upgradedAgentsIndex[agentIndex] = true
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, true}
 		}
 
 		ku.logger.Infof("Starting upgrade of agent nodes in pool identifier: %s, name: %s...\n",
 			*agentPool.Identifier, *agentPool.Name)
 
 		for _, vm := range *agentPool.AgentVMs {
-			ku.logger.Infof("Upgrading Agent VM: %s, pool name: %s\n", *vm.Name, *agentPool.Name)
-
 			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
+			upgradeVMs[agentIndex] = &UpgradeVM{*vm.Name, false}
+		}
 
-			err := upgradeAgentNode.DeleteNode(vm.Name)
-			if err != nil {
-				ku.logger.Infof("Error deleting agent VM: %s\n", *vm.Name)
-				return err
-			}
+		// Create an auxiliary node if hasn't been created yet.
+		// This node is used to take on the load from deleting nodes
+		addedNode := false
+		// In a normal mode of operation the actual number of VMs in the pool [len(upgradeVMs)] is equal to agentCount.
+		// However, if the upgrade failed in the middle, the actual number of VMs might be agentCount+1 to include auxiliary node.
+		// In the later case we don't create an extra node.
+		if agentCount > 0 && len(upgradeVMs) == agentCount {
```

Comment:
not sure we really need to be concerned about being asked to upgrade 0 node agent pools, but I guess it doesn't hurt.